clang warnings

Getting these warnings on FreeBSD with clang-40.


/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/NCAnalysisUGens/MedianSeparation.cpp:385:26: warning: comparison of unsigned expression >= 0 is always true [-Wtautological-compare]
            if ((bufnum1 >= 0) && (bufnum2 >= 0) ) {
                 ~~~~~~~ ^  ~
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/NCAnalysisUGens/MedianSeparation.cpp:385:44: warning: comparison of unsigned expression >= 0 is always true [-Wtautological-compare]
            if ((bufnum1 >= 0) && (bufnum2 >= 0) ) {
                                   ~~~~~~~ ^  ~
2 warnings generated.
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/PitchDetection/Qitch.cpp:194:53: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
        if (!((ampbufnum > world->mNumSndBufs) || ampbufnum<0)) {
                                                  ~~~~~~~~~^~
1 warning generated.
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/OteyPianoUGens/OteyPiano.cpp:223:3: warning: delete called on 'Hammer' that is abstract but has non-virtual destructor [-Wdelete-non-virtual-dtor]
  delete hammer;
  ^
1 warning generated.
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/RFWUGens/RFWUGens.cpp:186:51: warning: unsequenced modification and access to 'count' [-Wunsequenced]
        average = ((count * average) + *(in+i)) / ++count;
                    ~~~~~                         ^
1 warning generated.
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/BlackrainUGens/BlackrainUGens.cpp:250:14: warning: absolute value function 'fabsf' given an argument of type 'double' but has parameter of type 'float' which may cause truncation of value [-Wabsolute-value]
  float x1 = fabsf( input + _limit );
             ^
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/BlackrainUGens/BlackrainUGens.cpp:250:14: note: use function 'std::abs' instead
  float x1 = fabsf( input + _limit );
             ^~~~~
             std::abs
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/BlackrainUGens/BlackrainUGens.cpp:251:14: warning: absolute value function 'fabsf' given an argument of type 'double' but has parameter of type 'float' which may cause truncation of value [-Wabsolute-value]
  float x2 = fabsf( input - _limit );
             ^
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/BlackrainUGens/BlackrainUGens.cpp:251:14: note: use function 'std::abs' instead
  float x2 = fabsf( input - _limit );
             ^~~~~
             std::abs
2 warnings generated.
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/JoshUGens/JoshGrainUGens.cpp:6739:9: warning: 'GRAIN_BUF' macro redefined [-Wmacro-redefined]
#define GRAIN_BUF                                                                                                       \
        ^
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/JoshUGens/JoshGrainUGens.cpp:731:9: note: previous definition is here
#define GRAIN_BUF \
        ^
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/JoshUGens/JoshGrainUGens.cpp:6815:9: warning: 'GRAIN_LOOP_BODY_4' macro redefined [-Wmacro-redefined]
#define GRAIN_LOOP_BODY_4                                                                               \
        ^
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/JoshUGens/JoshGrainUGens.cpp:746:9: note: previous definition is here
#define GRAIN_LOOP_BODY_4 \
        ^
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/JoshUGens/JoshGrainUGens.cpp:6844:9: warning: 'GRAIN_LOOP_BODY_2' macro redefined [-Wmacro-redefined]
#define GRAIN_LOOP_BODY_2                                                               \
        ^
/usr/ports/audio/sc3-plugins/work/sc3-plugins-Version-3.9.0/source/JoshUGens/JoshGrainUGens.cpp:776:9: note: previous definition is here
#define GRAIN_LOOP_BODY_2 \
        ^

--- a/source/BlackrainUGens/BlackrainUGens.cpp
+++ b/source/BlackrainUGens/BlackrainUGens.cpp
@@ -246,9 +246,9 @@ void AmplitudeMod_next(AmplitudeMod* unit, int inNumSamples)
 }
 
 static inline float saturate(float input) {
-#define _limit 0.95
-  float x1 = fabsf( input + _limit );
-  float x2 = fabsf( input - _limit );
+  const float limit = 0.95f;
+  float x1 = fabsf( input + limit );
+  float x2 = fabsf( input - limit );
   return 0.5 * (x1 - x2);
 }
 

--- a/source/JoshUGens/JoshGrainUGens.cpp
+++ b/source/JoshUGens/JoshGrainUGens.cpp
@@ -6735,7 +6735,7 @@ inline float grain_in_at(Unit* unit, int index, int offset)
 		return IN0(index);
 }
 
-
+#undef GRAIN_BUF
 #define GRAIN_BUF													\
 	const SndBuf *buf;												\
 	if (bufnum >= world->mNumSndBufs) {								\
@@ -6812,6 +6812,7 @@ static inline bool getGrainWin(Unit * unit, float wintype, SndBuf *& window, con
 	return true;
 }
 
+#undef GRAIN_LOOP_BODY_4
 #define GRAIN_LOOP_BODY_4										\
 		float amp = y1 * y1;									\
 		phase = sc_gloop(phase, loopMax);						\
@@ -6841,6 +6842,7 @@ static inline bool getGrainWin(Unit * unit, float wintype, SndBuf *& window, con
 		y2 = y1;												\
 		y1 = y0;
 
+#undef GRAIN_LOOP_BODY_2
 #define GRAIN_LOOP_BODY_2								\
 		float amp = y1 * y1;							\
 		phase = sc_gloop(phase, loopMax);				\

--- a/source/NCAnalysisUGens/MedianSeparation.cpp
+++ b/source/NCAnalysisUGens/MedianSeparation.cpp
@@ -381,237 +381,232 @@ void MedianSeparation_next( MedianSeparation *unit, int inNumSamples ) {
             //printf("now recall %d %d \n",bufnum1,bufnum2);
             
             World *world = unit->mWorld;
-            
-            if ((bufnum1 >= 0) && (bufnum2 >= 0) ) { 
-                
-                SndBuf *buf1, *buf2;
-                
-                if (bufnum1 >= world->mNumSndBufs) { 
-                    int localBufNum = bufnum1 - world->mNumSndBufs; 
-                    Graph *parent = unit->mParent; 
-                    if(localBufNum <= parent->localBufNum) { 
-                        buf1 = parent->mLocalSndBufs + localBufNum; 
-                    } else { 
-                        buf1 = world->mSndBufs; 
-                    } 
-                } else { 
-                    buf1 = world->mSndBufs + bufnum1; 
-                } 
-                
-                LOCK_SNDBUF(buf1); 
-                
-                if (bufnum2 >= world->mNumSndBufs) { 
-                    int localBufNum = bufnum2 - world->mNumSndBufs; 
-                    Graph *parent = unit->mParent; 
-                    if(localBufNum <= parent->localBufNum) { 
-                        buf2 = parent->mLocalSndBufs + localBufNum; 
-                    } else { 
-                        buf2 = world->mSndBufs; 
-                    } 
-                } else { 
-                    buf2 = world->mSndBufs + bufnum2; 
-                } 
-                
-                LOCK_SNDBUF(buf2); 
-                
-                int numbins1 = (buf1->samples >> 1)  + 1; 
-                int numbins2 = (buf2->samples >> 1)  + 1; 
-                
-                //printf("check %d %d numbands %d further check %d %d \n",numbins1,numbins2,numbands,buf1->samples,buf2->samples);
-                
-                
-                if((numbins1 == numbands) && (numbins2 ==numbands)) {
-                    
-                    int hardorsoft = ZIN0(5); 
-                    
-                    //unit->hardorsoft_ = ZIN0(5); 
-                    //unit->p_ = ZIN0(6);
-                    
-                    //to magnitude and phase representation 
-                    SCPolarBuf *polar1 = ToPolarApx(buf1);
-                    SCPolarBuf *polar2 = ToPolarApx(buf2);
-                    
-                    SCPolar * data1 = polar1->bin; 
-                    SCPolar * data2 = polar2->bin; 
-                    
-                    //writing into the two output arrays of the buffers 
-                    
-                    
-                    //magsnow already pointing to write place in magnitudes
-                    //magsnow = mags + (midindex*numbands);
-                    
-                    //+mid+2, adding so no danger, just add 1
-                    int phaseindex = (unit->phaseposition_+1)%(mid+1); //midpoint is next one (about to be overwritten after increment below)
-                    
-                    float * phasesnow = phases + (phaseindex*numbands); 
-                    
-                    //dc, nyquist, phase to zero
-                    
-                    //buf1,polar 1 is harmonic, 2 is percussive
-                    
-                    //0 larger of horizontal and vertical is winner, or 1 more subtle blend 
-                    if(hardorsoft==0) {
-                        //hard
-                        
-                        //printf("hard calc \n");     
-                        
-                        if(horizontal[0]>vertical[0]) {
-                            
-                            polar1->dc = magsnow[0]; 
-                            polar2->dc = 0.f;
-                            
-                        } else {
-                            
-                            polar2->dc = magsnow[0]; 
-                            polar1->dc = 0.f;
-                        }
-                        
-                        
-                        if(horizontal[top]>vertical[top]) {
-                            
-                            polar1->nyq = magsnow[top]; 
-                            polar2->nyq = 0.f;
-                            
-                        } else {
-                            
-                            polar2->nyq = magsnow[top]; 
-                            polar1->nyq = 0.f;
-                        }
-                        
-                        int count = 0;  
-                        
-                        //setting 
-                        for (i=0; i<numbands-2; ++i) {
-                            
-                            int indexnow = i+1; 
-                            
-                            //if(i==20) {data1[i].mag=1024; data2[i].mag=1024;}
-                            
-                            //else {data1[i].mag = 0.0f; data2[i].mag = 0.f;}
-                            
-                            
-                            if(horizontal[indexnow]>vertical[indexnow]) {
-                                
-                                ++count; 
-                                
-                                data1[i].mag = magsnow[indexnow];
-                                data1[i].phase = phasesnow[indexnow];
-                                data2[i].mag = 0.0f;
-                                data2[i].phase = 0.0f; //phasesnow[i+1];
-                                
-                            } else {
-                                
-                                data2[i].mag = magsnow[indexnow];
-                                data2[i].phase = phasesnow[indexnow];
-                                data1[i].mag = 0.0f;
-                                data1[i].phase = 0.0f; //phasesnow[i+1];
-                                
-                            }
-                            
-                            //                        if(i<10) {
-                            //                            
-                            //                            printf("mag %d %f %f horiz %f vert %f further %f %f \n ",i, data1[i].mag,data2[i].mag,horizontal[indexnow],vertical[indexnow],polar1->bin[i].mag,polar2->bin[i].mag);
-                            //                            
-                            //                        }
-                            //                        
-                        }
-                        
-                        
-                        //printf("count %d \n",count);
-                        
-                        
-                    } else {
-                        
-                        
-                        //printf("I hope not!\n");
-                        
-                        //soft, Wiener filtering
-                        float pfactor = ZIN0(6); 
-                        
-                        float maskp, maskh, hp, pp, combine;
-                        
-                        //dc
-                        hp = powf(horizontal[0],pfactor);
-                        pp = powf(vertical[0],pfactor);
-                        
-                        combine = hp+pp; 
-                        
-                        maskh = 0.f; 
-                        maskp = 0.f; 
-                        
-                        //watch for zeroes
-                        if(combine>0.00000000001f) {
-                            maskh = hp/combine; 
-                            maskp = pp/combine; 
-                        }
-                        
-                        polar1->dc = magsnow[0] * maskh;
-                        polar2->dc = magsnow[0] * maskp;
-                        
-                        //nyquist
-                        hp = powf(horizontal[top],pfactor);
-                        pp = powf(vertical[top],pfactor);
-                        
-                        combine = hp+pp; 
-                        
-                        maskh = 0.f; 
-                        maskp = 0.f; 
-                        
-                        //watch for zeroes
-                        if(combine>0.00000000001f) {
-                            maskh = hp/combine; 
-                            maskp = pp/combine; 
-                        }
-                        
-                        polar1->nyq = magsnow[top] * maskh;
-                        polar2->nyq = magsnow[top] * maskp;
-                        
-                        //setting 
-                        for (i=0; i<numbands-2; ++i) {
-                            
-                            int indexnow = i+1; 
-                            
-                            hp = powf(horizontal[indexnow],pfactor);
-                            pp = powf(vertical[indexnow],pfactor);
-                            
-                            combine = hp+pp; 
-                            
-                            //if(i<5) {                       
-                            //     printf("mag %d %f %f %f %f %f \n",i ,horizontal[indexnow],vertical[indexnow],hp,pp,combine);
-                            //}
-                            
-                            maskh = 0.f; 
-                            maskp = 0.f; 
-                            
-                            //watch for zeroes
-                            if(combine>0.00000000001f) {
-                                maskh = hp/combine; 
-                                maskp = pp/combine; 
-                            }
-                            
-                            data1[i].mag = magsnow[indexnow] * maskh;
-                            data1[i].phase = phasesnow[indexnow];
-                            data2[i].mag = magsnow[indexnow] * maskp;
-                            data2[i].phase = phasesnow[indexnow];
-                            
-                        }
-                        
-                        
-                    }
-                    
-                    
-                    //printf("now output %d %d \n",bufnum1,bufnum2);
-                    
-                    //update output
-                    
-                    OUT0(0) = bufnum1; //2.f; //-1.f; //bufnum1; //-1;
-                    OUT0(1) = bufnum2; //-1;
-                    
+
+            SndBuf *buf1, *buf2;
+
+            if (bufnum1 >= world->mNumSndBufs) {
+                int localBufNum = bufnum1 - world->mNumSndBufs;
+                Graph *parent = unit->mParent;
+                if(localBufNum <= parent->localBufNum) {
+                    buf1 = parent->mLocalSndBufs + localBufNum;
+                } else {
+                    buf1 = world->mSndBufs;
                 }
-                
-                
-            } 
-            
+            } else {
+                buf1 = world->mSndBufs + bufnum1;
+            }
+
+            LOCK_SNDBUF(buf1);
+
+            if (bufnum2 >= world->mNumSndBufs) {
+                int localBufNum = bufnum2 - world->mNumSndBufs;
+                Graph *parent = unit->mParent;
+                if(localBufNum <= parent->localBufNum) {
+                    buf2 = parent->mLocalSndBufs + localBufNum;
+                } else {
+                    buf2 = world->mSndBufs;
+                }
+            } else {
+                buf2 = world->mSndBufs + bufnum2;
+            }
+
+            LOCK_SNDBUF(buf2);
+
+            int numbins1 = (buf1->samples >> 1)  + 1;
+            int numbins2 = (buf2->samples >> 1)  + 1;
+
+            //printf("check %d %d numbands %d further check %d %d \n",numbins1,numbins2,numbands,buf1->samples,buf2->samples);
+
+
+            if((numbins1 == numbands) && (numbins2 ==numbands)) {
+
+                int hardorsoft = ZIN0(5);
+
+                //unit->hardorsoft_ = ZIN0(5);
+                //unit->p_ = ZIN0(6);
+
+                //to magnitude and phase representation
+                SCPolarBuf *polar1 = ToPolarApx(buf1);
+                SCPolarBuf *polar2 = ToPolarApx(buf2);
+
+                SCPolar * data1 = polar1->bin;
+                SCPolar * data2 = polar2->bin;
+
+                //writing into the two output arrays of the buffers
+
+
+                //magsnow already pointing to write place in magnitudes
+                //magsnow = mags + (midindex*numbands);
+
+                //+mid+2, adding so no danger, just add 1
+                int phaseindex = (unit->phaseposition_+1)%(mid+1); //midpoint is next one (about to be overwritten after increment below)
+
+                float * phasesnow = phases + (phaseindex*numbands);
+
+                //dc, nyquist, phase to zero
+
+                //buf1,polar 1 is harmonic, 2 is percussive
+
+                //0 larger of horizontal and vertical is winner, or 1 more subtle blend
+                if(hardorsoft==0) {
+                    //hard
+
+                    //printf("hard calc \n");
+
+                    if(horizontal[0]>vertical[0]) {
+
+                        polar1->dc = magsnow[0];
+                        polar2->dc = 0.f;
+
+                    } else {
+
+                        polar2->dc = magsnow[0];
+                        polar1->dc = 0.f;
+                    }
+
+
+                    if(horizontal[top]>vertical[top]) {
+
+                        polar1->nyq = magsnow[top];
+                        polar2->nyq = 0.f;
+
+                    } else {
+
+                        polar2->nyq = magsnow[top];
+                        polar1->nyq = 0.f;
+                    }
+
+                    int count = 0;
+
+                    //setting
+                    for (i=0; i<numbands-2; ++i) {
+
+                        int indexnow = i+1;
+
+                        //if(i==20) {data1[i].mag=1024; data2[i].mag=1024;}
+
+                        //else {data1[i].mag = 0.0f; data2[i].mag = 0.f;}
+
+
+                        if(horizontal[indexnow]>vertical[indexnow]) {
+
+                            ++count;
+
+                            data1[i].mag = magsnow[indexnow];
+                            data1[i].phase = phasesnow[indexnow];
+                            data2[i].mag = 0.0f;
+                            data2[i].phase = 0.0f; //phasesnow[i+1];
+
+                        } else {
+
+                            data2[i].mag = magsnow[indexnow];
+                            data2[i].phase = phasesnow[indexnow];
+                            data1[i].mag = 0.0f;
+                            data1[i].phase = 0.0f; //phasesnow[i+1];
+
+                        }
+
+                        //                        if(i<10) {
+                        //
+                        //                            printf("mag %d %f %f horiz %f vert %f further %f %f \n ",i, data1[i].mag,data2[i].mag,horizontal[indexnow],vertical[indexnow],polar1->bin[i].mag,polar2->bin[i].mag);
+                        //
+                        //                        }
+                        //
+                    }
+
+
+                    //printf("count %d \n",count);
+
+
+                } else {
+
+
+                    //printf("I hope not!\n");
+
+                    //soft, Wiener filtering
+                    float pfactor = ZIN0(6);
+
+                    float maskp, maskh, hp, pp, combine;
+
+                    //dc
+                    hp = powf(horizontal[0],pfactor);
+                    pp = powf(vertical[0],pfactor);
+
+                    combine = hp+pp;
+
+                    maskh = 0.f;
+                    maskp = 0.f;
+
+                    //watch for zeroes
+                    if(combine>0.00000000001f) {
+                        maskh = hp/combine;
+                        maskp = pp/combine;
+                    }
+
+                    polar1->dc = magsnow[0] * maskh;
+                    polar2->dc = magsnow[0] * maskp;
+
+                    //nyquist
+                    hp = powf(horizontal[top],pfactor);
+                    pp = powf(vertical[top],pfactor);
+
+                    combine = hp+pp;
+
+                    maskh = 0.f;
+                    maskp = 0.f;
+
+                    //watch for zeroes
+                    if(combine>0.00000000001f) {
+                        maskh = hp/combine;
+                        maskp = pp/combine;
+                    }
+
+                    polar1->nyq = magsnow[top] * maskh;
+                    polar2->nyq = magsnow[top] * maskp;
+
+                    //setting
+                    for (i=0; i<numbands-2; ++i) {
+
+                        int indexnow = i+1;
+
+                        hp = powf(horizontal[indexnow],pfactor);
+                        pp = powf(vertical[indexnow],pfactor);
+
+                        combine = hp+pp;
+
+                        //if(i<5) {
+                        //     printf("mag %d %f %f %f %f %f \n",i ,horizontal[indexnow],vertical[indexnow],hp,pp,combine);
+                        //}
+
+                        maskh = 0.f;
+                        maskp = 0.f;
+
+                        //watch for zeroes
+                        if(combine>0.00000000001f) {
+                            maskh = hp/combine;
+                            maskp = pp/combine;
+                        }
+
+                        data1[i].mag = magsnow[indexnow] * maskh;
+                        data1[i].phase = phasesnow[indexnow];
+                        data2[i].mag = magsnow[indexnow] * maskp;
+                        data2[i].phase = phasesnow[indexnow];
+
+                    }
+
+
+                }
+
+
+                //printf("now output %d %d \n",bufnum1,bufnum2);
+
+                //update output
+
+                OUT0(0) = bufnum1; //2.f; //-1.f; //bufnum1; //-1;
+                OUT0(1) = bufnum2; //-1;
+
+            }
+
             unit->magnitudeposition_ = (unit->magnitudeposition_+1)%medsize; 
             unit->phaseposition_ = (unit->phaseposition_+1)%(mid+1); 
             

--- a/source/OteyPianoUGens/hammer.cpp
+++ b/source/OteyPianoUGens/hammer.cpp
@@ -38,10 +38,6 @@ StulovHammer :: StulovHammer(float f, float Fs, float m, float K, float p, float
 
 }
 
-Hammer :: ~Hammer() {
-}
-BanksHammer :: ~BanksHammer() {
-}
 void StulovHammer::trigger(float v){
   // this->v0 = v;
    this->v = v;

--- a/source/OteyPianoUGens/hammer.h
+++ b/source/OteyPianoUGens/hammer.h
@@ -29,7 +29,7 @@ struct UnitDelay{
 class Hammer {
 public:
   Hammer(float f, float Fs, float m, float K, float p, float Z, float alpha, float v0);
-  ~Hammer();
+  virtual ~Hammer() = default;
    void* operator new(size_t sz){
 		return RTAlloc(gWorld, sz);
 	}
@@ -55,7 +55,6 @@ class BanksHammer:public Hammer
 {
 public:
   BanksHammer(float f, float Fs, float m, float K, float p, float Z, float alpha, float v0);
-  ~BanksHammer();
   float load( float vin);
   void trigger(float v);
   float vh;
@@ -69,7 +68,6 @@ class StulovHammer:public Hammer
 {
 public:
   StulovHammer(float f, float Fs, float m, float K, float p, float Z, float alpha, float v0);
-  ~StulovHammer();
   float load( float vin);
   void trigger(float v);
   float upprev;

--- a/source/PitchDetection/Qitch.cpp
+++ b/source/PitchDetection/Qitch.cpp
@@ -191,7 +191,7 @@ void Qitch_Ctor(Qitch* unit)
 		unit->m_amps[i]= g_amps[i];
 	
 	uint32 ampbufnum = (uint32)ZIN0(4);
-	if (!((ampbufnum > world->mNumSndBufs) || ampbufnum<0)) {
+	if (ampbufnum < world->mNumSndBufs) {
 		SndBuf *buf2 = world->mSndBufs + ampbufnum; 
 		
 		bufsize = buf2->samples;

--- a/source/RFWUGens/RFWUGens.cpp
+++ b/source/RFWUGens/RFWUGens.cpp
@@ -183,7 +183,8 @@ void AverageOutput_next( AverageOutput *unit, int inNumSamples ) {
     }
 
 	for (i=0; i<inNumSamples; ++i) {
-        average = ((count * average) + *(in+i)) / ++count;
+        average = ((count * average) + *(in+i)) / (count + 1);
+        ++count;
         ZXP(out) = average;
 	}
 

